active/0021-transfer-files-over-mqtt.md

```diff
@@ -0,0 +1,199 @@
+
+Known cases of device-to-cloud file transfer:
+
+* [CAN bus](https://en.wikipedia.org/wiki/CAN_bus) data
+* Image taken by industry camera for Quality Assurance
+* Large data file collected from forklift
+* Video and audio data from truck cars, and video data captured by inbound unloading cameras
+* Vehicle real-time logging, telemetry, messaging
+* Upload collected ML logs
+
+Known cases of cloud-to-device file transfer:
+
+* Upload AI/ML models
+* Firmware upgrades
+
+Even though devices could already send binary data in MQTT packets, it is not trivial to guarantee reliable file transfer without clear expectations from client and server side.
+
+## Terminology
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [BCP 14](https://www.rfc-editor.org/bcp/bcp14) [[RFC2119](https://www.rfc-editor.org/rfc/rfc2119)] [[RFC8174](https://www.rfc-editor.org/rfc/rfc8174)] when, and only when, they appear in all capitals, as shown here.
+
+The following terms are used as described in [MQTT Version 5.0 Specification](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html):
+* Application Message
+* Server
+* Client
+* Topic
+* Topic Name
+* Topic Filter
+* MQTT Control Packet
+
+*At least once*: a message can be delivered many times, but cannot be lost
+
+## Requirements
+
+* The protocol MUST use only PUBLISH type of MQTT Control Packet
+* The protocol MUST support transfer of file segments
+* Server MUST be able to verify integrity of each file segment and of the whole file
+* Client MAY know total file size when initiating the transfer
+* Client MAY abort file transfer
+* Server MAY ask the client to pause file transfer
+* Server MAY ask the client to abort file transfer
+* The protocol MUST guarantee "At least once" delivery
+* Server MUST NOT support subscription on topics dedicated for file transfer
+
+## AWS IoT MQTT-based file delivery (reference design)
+
+As an example of existing implementation we can look at AWS IoT Core [which provides functionality](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery.html) to [deliver files to IoT devices](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery-in-devices.html):
+
+![](0021-assets/aws-mqtt-file-delivery.png)
+
+## Design
+
+### Overview
+
+* Files are split in segments, segments can be of arbitrary length
+* Client MUST generate UUID according to [RFC 4122](https://www.rfc-editor.org/rfc/rfc4122) for each file being transferred and use it as file Id in Topic Name
+* Broker MUST validate that provided UUID conforms to [RFC 4122](https://www.rfc-editor.org/rfc/rfc4122)
+* Client calculates sha256 checksum of the segment it's about to send and sends it as part of Topic Name
```

Comment:
We should make this optional.
yesterday we discussed that doing sha256 checksum is quite an expensive thing.